### ISSUE TITLE
Re-assigning acl_datacenter

### ISSUE THREAD

**Issue by cleung2010** [hashicorp/consul#1186]
Right now I have a multi-datacenter setup, call it dc1, dc2, dc3, each with 5 server nodes. I have `"acl_datacenter": "dc1"` on all the server nodes at the moment, but would like to switch it to, say dc2. I got several questions regarding this process:

1) What would be the correct procedure to do so without bumping into potential problems? 

2) Would changing the `acl_datacenter` value preserve existing ACL tokens? 

3) If there is a client node on dc3 that is using ACL tokens from dc1 (since its globally valid), would that token still be valid if `acl_datacenter` is then changed to dc2?

**Comment by armon** [comment]
@cleung2010 The ACL data is not replicated between datacenters so this is a very difficult transition. You'd want to export all the ACL tokens from DC1, switch `acl_datacenter` to DC2 and then restore all the tokens.

Otherwise, once you do the switch, DC2 will be a "blank slate" from the ACL perspective.

**Comment by cleung2010** [comment]
@armon I see. Is there a location on the filesystem where the ACLs are stored, or do I have to query it via the API and export like like so?

On a separate, but related question, if I were to decommission dc1, would I have to restore KV's too? What if they were already replicated to dc2 and dc3 with consul-replicate?

**Comment by armon** [comment]
@cleung2010 You'd need to use the APIs to export/import unfortunately. The file storage is a bit opaque to manipulate. With the KVs there is nothing to do if you've already replicated across. 

That is about it, just the KV and ACL data to move. Catalog is auto-managed anyways, Serf is gossiped around.

**Comment by cleung2010** [comment]
Thanks for the prompt reply and help! I'll close this issue and provide an update once it's done for future users that might be end up doing this.

**Comment by cleung2010** [comment]
@armon One last thing, what would be the correct procedure to migrate the ACLs? It is like so:

1. Export ACLs on dc1
2. Write exported ACLs to dc2
3. Change `acl_datacenter` on all server nodes from dc1 to dc2 (including dc1 itself?)
4. Shut down dc1 cluster
5. ???
6. Profit

On step 2, `acl_datacenter` on dc2 still points to dc1, so would there be a conflict there? Also, on step 3, once dc1's `acl_datacenter` is changed to dc2 would there be a conflict (although it wouldn't matter so much in this case since it's going to be taken down)?

**Comment by armon** [comment]
You'd need to swap steps 2 and 3, since you cannot write them to DC2 until it is the authority. DC2 would just end up forwarding to DC1. If the `acl_datacenter` conflicted you may end up with some very strange forwarding loops.

**Comment by cleung2010** [comment]
That would mean that there will me a time frame were no ACLs are valid since it will be in a blank slate state (thus blocking all requests to access the KV if its a deny-by-default setup). Is there a way to avoid this, or is the downtime imminent?

**Comment by cleung2010** [comment]
@armon You mentioned that dc2 would end up forwarding to dc1, are you talking about requests that needs ACL tokens for permissions, or the ACL tokens themselves? If it is the former, what would happen during an outrage on dc1? Would dc2 just not able to perform anything if its under a deny-by-default policy? Or does it have the ACLs cached locally so that accessing the KV store and service registration is still possible?

In other words, what is the actual flow for checking if the ACL is valid in clusters where it's not the `acl_datacenter`? Does dc2 (or dc3) asks dc1 every time a request is made to check whether the ACL token used is valid?

**Comment by cleung2010** [comment]
@armon If the dc1 was to be decommissioned, what would be the process of moving clients that are currently part of dc1 to dc2? Would that imply cluster merging issues and other weirdness along the way?

**Comment by armon** [comment]
You should just "leave" the old datacenter completely, wipe the client state, and "join" the new DC. Otherwise you will have all sorts of terrible problems. Migration between DCs was not an anticipated use case unfortunately.

**Comment by cleung2010** [comment]
What do you mean by wipe the client state?

**Comment by armon** [comment]
Anything in the `data_dir`

**Comment by cleung2010** [comment]
Got it, thanks for getting back to me.

**Comment by cleung2010** [comment]
@armon In the case of doing the `acl_datacenter` swap, is it safe to assume that the services that have been registered before in dc2 with with a particular ACL token will not disappear? Is it just that I can no longer use ACL to register new services?

What would happen if I disable ACLs on dc2 entirely? Would service still be able to register regardless of whether `?token` parameter is provided? Ideally I would want zero downtime on services that are registered to dc2. How would this ACL migration process affect services (whether on dc1 or dc2)?

**Comment by armon** [comment]
@cleung2010 Services will not disappear, that is correct. If ACLs are disabled then the `?token` parameter is just ignored as well. You can do that as well, simply disable ACLs, do the migration, re-enable and import the old tokens. The only window of time with an issue is between enabling ACLs and importing the old tokens.

**Comment by cleung2010** [comment]
@armon Cool! What do you mean by 'issues' during the time between enabling ACLs and importing the old tokens? Is it actions that require ACLs such as new service registration and KV queries? I think I can live with that temporarily as long as I am not taking down existing services.

**Comment by armon** [comment]
Any requests made to update the catalog will fail if the tokens aren't yet restored. Existing data will not be modified.

**Comment by cleung2010** [comment]
Got it. Any chance the import will mess up the CreateIndex or ModifyIndex and screw up the cluster, or is that handled gracefully?

**Comment by armon** [comment]
It *should* be okay, but I've never tried an ACL DC migration
